### PR TITLE
Code assist for Magic functions

### LINKED ISSUE #670: Code assist for magic functions
This is the issue (maybe epic?) for adding magic code assist to lightning.

https://www.figma.com/proto/xM7pTfdR0CVJsdgJb9Tvuy/Magic-Functions?page-id=0%3A1&node-id=20%3A1857&viewport=-1031%2C-246%2C0.71&scaling=min-zoom&starting-point-node-id=3%3A636

- Navigate to the "code view"
- Click "metadata explorer"
- Click back to "function docs"

Sub-issues
* #658 
* #657 
* #681

Dependencies:
* #621

### PR DESCRIPTION
**Pull request by josephjclark**
This PR is introduces code complete and supporting UI for magic functions.

It generally resolves #670

image

## Issues:

- #658
- #657
- #670

This has ended up all being one big PR - it just made more sense that way!

## Code complete

Monaco allows us to extend its code completion suggestions with whatever we want.

This PR adds a "completion provider". At a high level, this will look at the thing under the cursor and check the associated jsdoc for a custom `@lookup`  tag (which contains a jsonpath query).

If it finds one, it will execute that query against the metadata model for the loaded credential and return the results as suggestions.

# Job Editor

This PR contains several improvements to the Job Editor panel:

- Tabs to support multiple secondary panels
- Horizontal and vertical orientation (I'm never sure which is which!!)
- Collapsed mode to show a bigger code window 

## Metadata Panel

The metadata panel will:

- Tell the user whether magic is available
- Tell the user when metadata was generated
- Show the user what metadata was generated

## Docs Panel

The docs panel has been improved with a little indicator to show which functions have magic enabled

## Workflows Icon

The main Workflows icon has changed because I need to re-purpose the sparkle for magic stuff.

## General Architecture

Given a credential id, the Lightning client can call up to the Lightning Server to request metadata.

The server can use the CLI to do the generation (with caching).

Metadata is returned to the client as JSON. This then feeds into the Editor component (for code assist) and the new Metadata panel (for visualisation and information).

## Still to do here

* [x] ~~Report credential name (with link?)~~
* [x] Hook up to lightning metadata endpoint
* [x] Remove hard-coded metadata and dts
* [x] Loading states
* [x] Resizable panels!! 
* [x] ~~Better error states~~
* [x] ~~More drama when metadata is loaded (I want to colour the tab icon)~~
* [x] Better help panel down at the bottom (persistence, better affordance)
* [x] ~~Better presentation of types inside the metadata explorer (they float horribly right now, it's a difficult layout problem)~~

## Further Work

* We may need language-specific insertion templates - or least options. When inserting a dhis2 id, it makes sense to comment in the label:
 ```
attribute: "ihn1wb9eho8" /*Catchment area*/
 ```
But this isn't true for all insertions. We either need to work out a ruleset (add a comment if the label is different to the value) or give adaptors a means to control this (perhaps @lookupPath and @lookupValue).

Update: I think dhis2 have changed their sandbox data so that the labels and names are better defined now. Certianly last time I ran metadata this problem seemed to go away.

### CODE REVIEW

**Comment by josephjclark** [comment]
This could benefit from a review.

The resize stuff is sitting in #699, I'd like to merge if I can fix it.

The lightning side of this needs looking at. I could use some help getting the credential name down into the React components!

**Comment by josephjclark** [comment]
Notes on integrating with Stu's backend service:

* Clientside things look lovely
* Looks like you have to have run a job to install an adaptor before the metadata service will run. I guess this is a consequence of https://github.com/OpenFn/kit/issues/192 ?

**Comment by stuartc** [comment]
> Notes on integrating with Stu's backend service:
> 
> * Clientside things look lovely
> * Looks like you have to have run a job to install an adaptor before the metadata service will run. I guess this is a consequence of [CLI metadata: allow autoinstall kit#192](https://github.com/OpenFn/kit/issues/192) ?

Ah yeah that makes sense, don't worry about the CLI doing auto install - we use a global install strategy for lightning; the `AdaptorService` can do autoinstalls if asked. I'll add that.

> * Metadata service has a minimum CLI version compatibility - and as we make changes we may continue to have version stuff. Should we have some kind of version checker? Shouldn't matter for prod but it'll help devs

I agree, as much as I want to make it backwards compatible - I think I will make it check for a minimum version for now. I did want to make it check for the existence of the CLI, I should do that and check for a version number as well - that seem good enough for now?

> * My credential is `{ configuration: { username... } }`, not ` { username:... }`. That's a bit of a surprise

Ah right, yeah thats my bad - I pass the credential body in _as is_, so technically it's not 'state'. I can fix that.

> * Bug: the loading symbol in the editor doesn't layout properly any more

Hmm, yeah not sure about how that happened. :shrug: Ideas?

> * Reminder: We haven't really discussed core language definitions (es5.d.ts)

Right, please fill me in on that - it's been a long time since I dealt with loading in base types.

> * Still seeing problems with the editor occasionally disappearing on resize (did I re-introduce this? The old solution was never very good...)

Not sure, couldn't recreate it in my very limited testing.

> * Currently seeing some issues on dhis2

...Go on...

**Comment by josephjclark** [comment]
> I see quite a few errors from jsdelivr when the editor tries to get ^1.7.5 for common 

Oh is that what it is! I saw the error last night but didn't really grok it. Pretty frustrating.

That fix looks fine for the moment, but this is yet another reason to take some time out and get this issue sorted: https://github.com/OpenFn/kit/issues/197

**Comment by josephjclark** [comment]
@stuartc A lot of those comments are for me, not for you :) Just notes for what I need to fix. I'll be on it today. Thanks for the backend fixes though!

**Comment by stuartc** [comment]
@josephjclark I just added another commit, doesn't really affect the FE, but just FYI another possible error message now would be: `"too_many_processes"`. 

Thats from the rate limiter for CLI calls... I'll eventually wrap that in a retry mechanism - but we at least have _some_ kind of protection from too many calls that make child processes.

**Comment by stuartc** [comment]
> > I see quite a few errors from jsdelivr when the editor tries to get ^1.7.5 for common
> 
> Oh is that what it is! I saw the error last night but didn't really grok it. Pretty frustrating.
> 
> That fix looks fine for the moment, but this is yet another reason to take some time out and get this issue sorted: [OpenFn/kit#197](https://github.com/OpenFn/kit/issues/197)

I'm all for moving forward on that issue. I'll make some time to understand all the ins and outs of the proposals/considerations in the issue.

**Comment by josephjclark** [comment]
Sooo I think this is ready for a little QA! @amberrignell maybe you'd like to take a look?

Magic functions are working, the backend is in place, everything looks more or less OK to me.

One big thing to bear in mind: the adaptor will not auto install (yet). You may need to run a job before metadata will load.

File: assets/css/app.css [hunk 1]
```diff
@@ -106,6 +106,10 @@
   animation: 0.2s ease-out 0s normal forwards 1 fade-out-keys;
 }
 
+.flex-col-reverse {
+  flex-direction: column-reverse;
+}
```
Comment (stuartc) [inline assets/css/app.css:111]: @josephjclark isn't this already in tailwind?

**Review by stuartc — COMMENTED** [review]
Awesome work! It would be helpful to get a sense on what needs tidying up now, and or is possibly a bit brittle. With all these changes I'm sure there is a bit of that.

Oh, we need the CHANGELOG updated please! :smile:

File: assets/css/app.css (hunk 1 above)
Comment (taylordowns2000) [reply assets/css/app.css]: removing as this doesn't appear to be used anywhere in the app :-)

**Pull request merged**